File: scripts/readme-to-pdf.ts
import fs from 'fs'
import path from 'path'
import { jsPDF } from 'jspdf'
import autoTable from 'jspdf-autotable'

const inputPath = path.join(__dirname, '../README_REPO.md')
const outputPath = path.join(__dirname, '../README_REPO.pdf')

const md = fs.readFileSync(inputPath, 'utf-8')

const doc = new jsPDF({ unit: 'pt', format: 'a4' })
const margin = 36
let y = margin
const lineHeight = 16
const maxWidth = 520

function addTextBlock(text: string) {
  const lines: string[] = doc.splitTextToSize(text, maxWidth)
  lines.forEach((line: string) => {
    if (y > 780 - margin) {
      doc.addPage()
      y = margin
    }
    doc.text(line, margin, y)
    y += lineHeight
  })
  y += 4
}

function addTable(headers: string[], rows: string[][]) {
  autoTable(doc, {
    head: [headers],
    body: rows,
    startY: y,
    margin: { left: margin, right: margin },
    theme: 'grid',
    styles: { fontSize: 9 },
    headStyles: { fillColor: [80, 80, 160] },
    didDrawPage: (data) => { y = (data.cursor?.y ?? y) + 10 }
  })
}

const lines = md.split(/\r?\n/)
for (let i = 0; i < lines.length; i++) {
  let line = lines[i]
  if (line.startsWith('# ')) {
    doc.setFontSize(20)
    doc.setFont('helvetica', 'bold')
    addTextBlock(line.replace('# ', ''))
    doc.setFontSize(12)
    doc.setFont('helvetica', 'normal')
  } else if (line.startsWith('## ')) {
    doc.setFontSize(15)
    doc.setFont('helvetica', 'bold')
    addTextBlock(line.replace('## ', ''))
    doc.setFontSize(12)
    doc.setFont('helvetica', 'normal')
  } else if (line.startsWith('|') && lines[i+1] && lines[i+1].startsWith('|')) {
    // Markdown table
    const header = line.split('|').map(s => s.trim()).filter(Boolean)
    const rows = []
    i++ // skip header separator
    while (lines[i+1] && lines[i+1].startsWith('|')) {
      i++
      const row = lines[i].split('|').map(s => s.trim()).filter(Boolean)
      rows.push(row)
    }
    addTable(header, rows)
  } else if (line.startsWith('```')) {
    // Code block
    let code = ''
    i++
    while (i < lines.length && !lines[i].startsWith('```')) {
      code += lines[i] + '\n'
      i++
    }
    doc.setFont('courier', 'normal')
    doc.setFontSize(10)
    addTextBlock(code.trim())
    doc.setFont('helvetica', 'normal')
    doc.setFontSize(12)
  } else if (line.trim() !== '') {
    addTextBlock(line)
  } else {
    y += 6
  }
}

doc.save(outputPath)
console.log('PDF created:', outputPath)

File: scripts/readme_to_pdf.py
import re
from fpdf import FPDF

# Read the markdown file
with open(r"C:\Users\nitis\OneDrive\Desktop\Astro\README_REPO.md", "r", encoding="utf-8") as f:
    content = f.read()

# Create PDF
pdf = FPDF()
pdf.set_auto_page_break(auto=True, margin=15)
pdf.add_page()

# Add a Unicode font (DejaVu is bundled with fpdf2)
pdf.add_font("DejaVu", "", "DejaVuSans.ttf", uni=True)
pdf.add_font("DejaVu", "B", "DejaVuSans-Bold.ttf", uni=True)

lines = content.split("\n")
in_code_block = False
in_table = False
table_rows = []

i = 0
while i < len(lines):
    line = lines[i]
    
    # Code block handling
    if line.startswith("```"):
        if in_code_block:
            in_code_block = False
        else:
            in_code_block = True
        i += 1
        continue
    
    if in_code_block:
        pdf.set_font("Courier", size=9)
        pdf.set_text_color(60, 60, 60)
        pdf.multi_cell(0, 5, line)
        pdf.set_text_color(0, 0, 0)
        i += 1
        continue
    
    # Table handling
    if line.startswith("|"):
        if not in_table:
            in_table = True
            table_rows = []
        # Skip separator row
        if re.match(r"^\|[\s\-:|]+\|$", line):
            i += 1
            continue
        cells = [c.strip() for c in line.split("|")[1:-1]]
        table_rows.append(cells)
        i += 1
        continue
    elif in_table:
        # End of table, render it
        in_table = False
        if table_rows:
            pdf.set_font("DejaVu", size=9)
            col_count = len(table_rows[0]) if table_rows else 1
            col_width = (pdf.w - 20) / col_count
            for row_idx, row in enumerate(table_rows):
                for cell in row:
                    if row_idx == 0:
                        pdf.set_font("DejaVu", "B", 9)
                        pdf.set_fill_color(80, 80, 160)
                        pdf.set_text_color(255, 255, 255)
                    else:
                        pdf.set_font("DejaVu", "", 9)
                        pdf.set_fill_color(245, 245, 245)
                        pdf.set_text_color(0, 0, 0)
                    pdf.cell(col_width, 7, cell[:40], border=1, fill=True)
                pdf.ln()
            pdf.set_text_color(0, 0, 0)
            pdf.ln(3)
            table_rows = []
    
    # Headers
    if line.startswith("# "):
        pdf.set_font("DejaVu", "B", 20)
        pdf.ln(5)
        pdf.multi_cell(0, 10, line[2:])
        pdf.ln(3)
    elif line.startswith("## "):
        pdf.set_font("DejaVu", "B", 15)
        pdf.ln(4)
        pdf.multi_cell(0, 8, line[3:])
        pdf.ln(2)
    elif line.startswith("### "):
        pdf.set_font("DejaVu", "B", 12)
        pdf.ln(3)
        pdf.multi_cell(0, 7, line[4:])
        pdf.ln(2)
    elif line.startswith("- "):
        pdf.set_font("DejaVu", "", 11)
        pdf.multi_cell(0, 6, "  • " + line[2:])
    elif line.strip():
        pdf.set_font("DejaVu", "", 11)
        pdf.multi_cell(0, 6, line)
    else:
        pdf.ln(3)
    
    i += 1

# Handle any remaining table
if in_table and table_rows:
    pdf.set_font("DejaVu", size=9)
    col_count = len(table_rows[0]) if table_rows else 1
    col_width = (pdf.w - 20) / col_count
    for row_idx, row in enumerate(table_rows):
        for cell in row:
            if row_idx == 0:
                pdf.set_font("DejaVu", "B", 9)
                pdf.set_fill_color(80, 80, 160)
                pdf.set_text_color(255, 255, 255)
            else:
                pdf.set_font("DejaVu", "", 9)
                pdf.set_fill_color(245, 245, 245)
                pdf.set_text_color(0, 0, 0)
            pdf.cell(col_width, 7, cell[:40], border=1, fill=True)
        pdf.ln()
    pdf.set_text_color(0, 0, 0)

# Save PDF
output_path = r"C:\Users\nitis\OneDrive\Desktop\Astro\README_REPO.pdf"
pdf.output(output_path)
print(f"PDF created: {output_path}")
